### PR TITLE
CPVPA workstream: Create ScalingPolicy duck-type interface

### PR DESCRIPTION
**Pull request by justinsb**
As the first step towards support other VPA policies via duck-typing,
we abstract out a ScalingPolicy interface.

The API type directly implements this interface.

### CODE REVIEW

**Comment by bskiba** [comment]
@kgolab and/or me will take a look hopefully next week.

File: vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta1/duck_types.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+			}
+			if containerPolicies[i].ContainerName == DefaultContainerResourcePolicy {
+				defaultPolicy = &containerPolicies[i]
+			}
+		}
+	}
+	return defaultPolicy
+}
+
+// GetContainerResourcePolicy returns the ContainerResourcePolicy for a given policy
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta1/duck_types.go]: It looks like we don't actually need this anymore and could as well switch to use ScalingPolicy interface everywhere which would be a cleaner option.

File: vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta1/duck_types.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	return v.Spec.Selector
+}
+
+func (v *VerticalPodAutoscaler) GetRecommendation() *RecommendedPodResources {
+	return v.Status.Recommendation
+}
+
+func (v *VerticalPodAutoscaler) GetContainerResourcePolicy(containerName string) *ContainerResourcePolicy {
+	var defaultPolicy *ContainerResourcePolicy
+	if v != nil && v.Spec.ResourcePolicy != nil {
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta1/duck_types.go]: My go is rusty but do we need to check if v != nil?

File: vertical-pod-autoscaler/pkg/recommender/model/cluster.go [hunk 3]
```diff
@@ -227,7 +227,7 @@ func (cluster *ClusterState) AddOrUpdateVpa(apiObject *vpa_types.VerticalPodAuto
 	}
 	vpa.Conditions = conditionsMap
 	vpa.Recommendation = currentRecommendation
-	vpa.ResourcePolicy = apiObject.Spec.ResourcePolicy
+	vpa.ScalingPolicy = apiObject.DeepCopy()
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/recommender/model/cluster.go]: I wonder - the ScalingPolicy contains Recommendation too (via GetRecommendation() method); what is the relation between this one & vpa.Recommendation?

**Comment by kgolab** [comment]
Justin, thanks for the PR.

Except for the smaller comments I have one larger request - could you please add (writable) Conditions to the ScalingPolicy and restore passing them to all the extension points where you've removed them (for example UpdatePriorityCalculator or RecommendationProcessor)?

**Comment by mwielgus** [comment]
@justinsb  Ping :). Please respond to the comments.

**Comment by mwielgus** [comment]
@justinsb Second Ping :). Please respond to the comments.

**Comment by mwielgus** [comment]
@justinsb Please respond to the comments. Otherwise this PR will be closed as inactive.

**Comment by mwielgus** [comment]
Closing as inactive. Please reopen once you will be able to continue work on this.

**Pull request closed without merging**